End to end testing demonstration using mocking

ayoung012 opened the pull request:
An end to end test for the SingleRandomAgent. Provide agent with basic step 0 percepts with a single shop, and assert the agent 'randomly' picks the only shop and verifies a goto action for the shop is submitted to the server.

This was far from simple to write. I was crawling examples of internal SARL tests and dealing with all sorts of weird and wacky limitations of mocking in Mockito.

Mockito works best if mocked objects are passed in. To this end, I created three seperate Initialize definitions to cover providing a skill direct from boot, providing a config directory, and providing no Params for interactive config selection. The test uses the Initialize provided with the skill.

I think this will be incredibly useful to extend with more complex scenarios and agents.

Further improvements made:
- Allow test to advance scenario by exactly one step, then make assertions about agent actions

ayoung012 committed:
allow injection of S_MassimTalking into SingleRandomAgent and EnvironmentInterface into S_MassimTalking

ayoung012 committed:
Basic e2e test for SingleRandomAgent

ssardina commented:
wow, this looks great, but I am trying to understand what exactly it is. It would be kind of a "unit test"?

wow I just looked a bit more, it seems you can use this WITHOUT running the server!! is this correct? I trust you, but I am very curious

ssardina commented on src/main/sarl/au/edu/rmit/agtgrp/agtcity/sarl/mw/agents/SingleRandomAgent.sarl:
```diff
@@ -106,44 +105,63 @@ agent SingleRandomAgent {
+	 * 1. FIRST, collect the directory where the server configuration file eiRegistrationFile is located
+	 * If none is given as CLI argument, search inside conf/ folder and make user select one
+	 */
+	on Initialize [ occurrence.parameters.length > 0 && occurrence.parameters.get(0) instanceof S_MassimTalking ] {
```
This is cool. So, if the agent receives the S_Massimtalking, then it will load it?

ayoung012 replied:
Yes. It is not really in the spirit of testing to provide a constructor specifically for tests, maybe we should always be passing skills as parameters?

ssardina replied:
Nice, I am reading and understanding... very sophisticated.. you clearly master SARL better than me!

ayoung012 replied:
The robustness of the sarl constructs is helping. I have not run into any major issues with taking the Mockito examples for java and bringing them to SARL

ayoung012 replied:
You see here that I have mocked EnvironmentInterface, the java gateway to the massim server. I have yet to figure out how to mock S_MassimTalking, as a skill is more complex than a simple java class when being handled by the SARL runtime engine.

srodriguez replied:
@ayoung012 Nice work! Happy to chat with you about testing options!

ayoung012 replied:
@srodriguez sounds great, lets do it

ayoung012 commented:
Yes, running the server is not required, production of percepts handled by thenAnswer. Would be nice to have more granular control within the unit test. Eg 
whenAgentsReceivePerceptsForStep(4)

ayoung012 commented:
> wow, this looks great, but I am trying to understand what exactly it is. It would be kind of a "unit test"?
> 
> wow I just looked a bit more, it seems you can use this WITHOUT running the server!! is this correct? I trust you, but I am very curious

I would not call this a 'unit test' as we do not dial in on one part of the agent in particular. I would call it an end-to-end test or a behaviour driven test

ayoung012 committed:
extract test setup to generic step files, clear and concise tests

ayoung012 commented:
I have made the actual junit 'tests' much more concise by extracting some generic patterns into helper classes.

I also implemented total control of what scenario steps to 'mock' send to the agent within the test. This allows for tests like:
givenSarlAgent("agent1")
givenSarlAgent("agent2")
whenAgentsPerceiveStep(1)
thenActionsPerformed(2)
...

joshuahansen commented on src/test/sarl/au/edu/rmit/agtgrp/agtcity/sarl/mw/SingleRandomAgentTest.sarl:
```diff
@@ -74,125 +61,115 @@ import au.edu.rmit.agtgrp.agtcity.sarl.mw.skills.S_MassimTalking
+	def beforeEach(extension rc : SreRunContext) : void {
+		ctx = rc
+		/* Initialise a fresh SARL environment for each test. */
+        kern = ctx.setupTheSreKernel(Level::FINEST, LogTrackingMode::STANDARD_LOGGER_BEHAVIOR)
```
- [ ] Fix indent

ayoung012 replied:
dcbbcdc

joshuahansen reviewed (changes requested):
Few minor changes to clean up

ayoung012 committed:
example agent: terminate agent on end of scenario

ayoung012 committed:
indicate test completion by queueing the final scenario step.

Also use Logging interface, cleanup

ayoung012 commented on src/test/sarl/au/edu/rmit/agtgrp/agtcity/sarl/mw/ScenarioStepper.sarl:
```diff
@@ -0,0 +1,64 @@
+		}
+		/* Timeout, no steps were given for us to send. Send end game. */
+		System.err.println("Expecting steps for " + callingAgent + " to send, had none, timeout")
+		//return step.put(callingAgent, stepLast())
```
dcbbcdc

ayoung012 commented on src/test/sarl/au/edu/rmit/agtgrp/agtcity/sarl/mw/MassimSteps.sarl:
```diff
@@ -0,0 +1,110 @@
+	 * !!! REQUIRES SPECIFIC MOCKITO CONFIGURATION !!!
+	 * src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker
+	 * 
+	 * Verify that a number of actions have been sent to the given {@link EnvironmentInterface}.
```
dcbbcdc

ayoung012 commented on src/test/sarl/au/edu/rmit/agtgrp/agtcity/sarl/mw/MassimSteps.sarl:
```diff
@@ -0,0 +1,110 @@
+	 * Mock some {@code EnvironmentInterface} methods to indicate that the massim server is
+	 * in a given state with some given agents.
+	 *
+	 * @param ei mocked
```
dcbbcdc

ayoung012 committed:
Merge remote-tracking branch 'origin/master' into feature/agent-testing

ayoung012 committed:
update to sarl 0.12

ayoung012 committed:
update dependency dependency exlusions for sarl 0.12

Better, but seems some eclipse dependencies still snuck in

ayoung012 committed:
improve e2e testing API:

- allow more generic args to be passed to agent
- allow timeout duration override
- allow multiple action assertions in a single test

ayoung012 committed:
bundle a test jar to facilitate writing tests on sarl agent city projects

ayoung012 committed:
small cleanup of noisy logs, Pure functions

ayoung012 committed:
test jar assembly conf

ayoung012 committed:
allow provision of a default timeout step to send
